### ISSUE TITLE
Question about validation gradient summation in ghost computation for Conv1D and Conv2D layers.

### ISSUE THREAD

**Issue by MarkWang1196** [Jiachen-T-Wang/GhostSuite#8]
Hi, thank you for this excellent work on efficient gradient dot-product computation!

I have a question about the implementation of ghost computation in the _compute_Conv1D_dot_product function (line 718 in ghostEngines/supported_layers_grad_samplers_dotprod.py).

### Current Implementation
The current code computes the validation gradient by first summing over the validation batch, then computing the outer products:
``` python
if layer.use_ghost_computation:
    # Sum over the validation batch dimension first
    A_val_sum = torch.sum(A_val_f, dim=0)  # [t, d]
    B_val_sum = torch.sum(B_val_f, dim=0)  # [t, p]
    
    # Then compute outer products with training samples
    AA = torch.matmul(A_val_sum.unsqueeze(0), A_train_f.transpose(-1, -2))  # [b, t, t]
    BB = torch.matmul(B_val_sum.unsqueeze(0), B_train_f.transpose(-1, -2))  # [b, t, t]
    
    # Final dot product
    layer.weight.grad_dot_prod = torch.sum((AA * BB).to(accum_dtype), dim=[1, 2])  # [b]
```
This computes:
``` python
dot_prod[i] = sum((A_train[i]^T @ A_val_sum) ⊙ (B_train[i]^T @ B_val_sum))
```
where` A_val_sum = Σⱼ A_val[j] `and` B_val_sum = Σⱼ B_val[j]`.

### Question
However, mathematically, the gradient dot-product should be:

``` python
dot_prod[i] = <grad_train[i], Σⱼ grad_val[j]>
            = Σⱼ <grad_train[i], grad_val[j]>
            = Σⱼ sum((A_train[i]^T @ A_val[j]) ⊙ (B_train[i]^T @ B_val[j]))
``` 
The issue is that:
``` python
(A_train^T @ (Σⱼ A_val[j])) ⊙ (B_train^T @ (Σⱼ B_val[j]))
≠ Σⱼ ((A_train^T @ A_val[j]) ⊙ (B_train^T @ B_val[j]))
``` 
This is because when we expand the left side:
``` python
(A_train^T @ (A_val[1] + A_val[2])) ⊙ (B_train^T @ (B_val[1] + B_val[2]))
= (A_train^T @ A_val[1] + A_train^T @ A_val[2]) ⊙ (B_train^T @ B_val[1] + B_train^T @ B_val[2])
``` 
This produces cross terms that shouldn't exist:
``` python
= (A_train^T @ A_val[1]) ⊙ (B_train^T @ B_val[1])  ✓
  + (A_train^T @ A_val[1]) ⊙ (B_train^T @ B_val[2])  ✗ (extra)
  + (A_train^T @ A_val[2]) ⊙ (B_train^T @ B_val[1])  ✗ (extra)
  + (A_train^T @ A_val[2]) ⊙ (B_train^T @ B_val[2])  ✓
``` 

Once again, thank you very much for open-sourcing your code. If any of my observations above are incorrect, I would greatly appreciate your indulgence.
